core/src/test/java/org/elasticsearch/NamingConventionTests.java

```diff
@@ -143,39 +151,57 @@ public FileVisitResult postVisitDirectory(Path dir, IOException exc) throws IOEx
         }
         assertTrue(missingSuffix.remove(WrongName.class));
         assertTrue(missingSuffix.remove(WrongNameTheSecond.class));
+        assertTrue(notRunnable.remove(DummyAbstractTests.class));
+        assertTrue(notRunnable.remove(DummyInterfaceTests.class));
+        assertTrue(innerClasses.remove(InnerTests.class));
         assertTrue(notImplementing.remove(NotImplementingTests.class));
         assertTrue(notImplementing.remove(NotImplementingTest.class));
         assertTrue(pureUnitTest.remove(PlainUnit.class));
         assertTrue(pureUnitTest.remove(PlainUnitTheSecond.class));
 
         String classesToSubclass = Joiner.on(',').join(
-                ESTestCase.class.getSimpleName(),
-                ESTestCase.class.getSimpleName(),
-                ESTokenStreamTestCase.class.getSimpleName(),
-                LuceneTestCase.class.getSimpleName());
+            ESTestCase.class.getSimpleName(),
+            ESTestCase.class.getSimpleName(),
+            ESTokenStreamTestCase.class.getSimpleName(),
+            LuceneTestCase.class.getSimpleName());
         assertTrue("Not all subclasses of " + ESTestCase.class.getSimpleName() +
-                        " match the naming convention. Concrete classes must end with [Test|Tests]: " + missingSuffix.toString(),
-                missingSuffix.isEmpty());
-        assertTrue("Pure Unit-Test found must subclass one of [" + classesToSubclass +"] " + pureUnitTest.toString(),
-                pureUnitTest.isEmpty());
-        assertTrue("Classes ending with Test|Tests] must subclass [" + classesToSubclass +"] " + notImplementing.toString(),
-                notImplementing.isEmpty());
-        assertTrue("Subclasses of ESIntegTestCase should end with IT as they are integration tests: " + integTestsInDisguise, integTestsInDisguise.isEmpty());
+                " match the naming convention. Concrete classes must end with [Test|Tests]:\n" + listClasses(missingSuffix),
+            missingSuffix.isEmpty());
+        assertTrue("Classes ending with [Test|Tests] are abstract or interfaces:\n" + listClasses(notRunnable),
+            notRunnable.isEmpty());
+        assertTrue("Found inner classes that are tests, which are excluded from the test runner:\n" + listClasses(innerClasses),
+            innerClasses.isEmpty());
+        assertTrue("Pure Unit-Test found must subclass one of [" + classesToSubclass +"]:\n" + listClasses(pureUnitTest),
+            pureUnitTest.isEmpty());
+        assertTrue("Classes ending with Test|Tests] must subclass [" + classesToSubclass +"]:\n" + listClasses(notImplementing),
+            notImplementing.isEmpty());
+        assertTrue("Subclasses of ESIntegTestCase should end with IT as they are integration tests:\n" + listClasses(integTestsInDisguise), integTestsInDisguise.isEmpty());
+    }
+
+    static String listClasses(Set<Class> classes) {
```

Comment:
This is one of those cases where I prefer Guava's `Joiner.on("\n").join(classes)`